bookdown/05-extending/05-extending-mlr3pipelines.Rmd

```diff
@@ -1,5 +1,470 @@
+
+When we inherit from `PipeOpTaskPreproc`, it sets the `input` and `output` `data.table`s for us to only accept a single `Task`. The only thing we do during `initialize()` is therefore to set an `id` (which can optionally be changed by the user).
+
+The complete `PipeOpDropNA` can therefore be written as follows. Note that it inherits from `PipeOpTaskPreproc`, unlike the `PipeOpCopyTwo` example from above:
+
+```{r 05-extending-021}
+PipeOpDropNA = R6::R6Class("PipeOpDropNA",
+  inherit = PipeOpTaskPreproc,
+  public = list(
+    initialize = function(id = "drop.na") {
+      super$initialize(id)
+    },
+
+    train_task = function(task) {
+      self$state = list()
+      featuredata = task$data(cols = task$feature_names)
+      exclude = apply(is.na(featuredata), 1, any)
+      task$filter(task$row_ids[!exclude])
+    },
+
+    predict_task = function(task) {
+      # nothing to be done
+      task
+    }
+  )
+)
+```
+
+To test this `PipeOp`, we create a small task with missing values:
+```{r 05-extending-022}
+smalliris = iris[(1:5) * 30, ]
+smalliris[1, 1] = NA
+smalliris[2, 2] = NA
+sitask = TaskClassif$new("smalliris", as_data_backend(smalliris), "Species")
+print(sitask$data())
+```
+
+We test this by feeding it to a new `Graph` that uses `PipeOpDropNA`.
+```{r 05-extending-023}
+gr = Graph$new()
+gr$add_pipeop(PipeOpDropNA$new())
+
+filtered_task = gr$train(sitask)[[1]]
+print(filtered_task$data())
+```
+
+
+#### Example: `PipeOpScaleAlways`
+
+An often-applied preprocessing step is to simply **center** and/or **scale** the data to mean $0$ and standard deviation $1$.
+This fits the `PipeOpTaskPreproc` pattern quite well.
+ Because it always replaces all columns that it operates on, and does not require any information about the task's target, it only needs to overload the `train_dt` and `predict_dt` functions.
+ This saves some boilerplate-code from getting the correct feature columns out of the task, and replacing them after modification.
+
+Because scaling only makes sense on numeric features, we want to instruct `PipeOpTaskPreproc` to give us only these numeric columns.
+We do this by overloading the `select_cols` function: It is called by the class to determine which columns to give to `train_dt` and `predict_dt`.
+Its input is the `Task` that is being transformed, and it should return a `character` vector of all features to work with. When it is not overloaded, it uses all columns; instead, we will set it to only give us numeric columns.
+Because the `levels()` of the data table given to `train_dt` and `predict_dt` may be different from the levels `task`'s levels, these functions must also take a `levels` argument that is a named list of column names indicating their levels. When working with numeric data, this argument can be ignored, but it should be used instead of `levels(dt[[column]])` for factorial or character columns.
```

Comment:
```suggestion
Because the `levels()` of the data table given to `train_dt` and `predict_dt` may be different from the levels `task`'s levels, these functions must also take a `levels` argument that is a named list of column names indicating their levels. 
When working with numeric data, this argument can be ignored, but it should be used instead of `levels(dt[[column]])` for factorial or character columns.
```